lang_regexp/parsing/Lexer.mll

```diff
@@ -0,0 +1,695 @@
+(*
+   Lexer for the regexp parser.
+
+   See Parse.mli for a list of to-dos.
+*)
+{
+open AST
+open Conf
+open Parser
+
+let loc lexbuf : loc =
+  let tok = Parse_info.tokinfo lexbuf in
+  (tok, tok)
+
+let int = int_of_string
+
+(*
+   Decode a well-formed UTF-8 string into a Unicode code point.
+*)
+let decode s =
+  let len = String.length s in
+  if len <= 0 then
+    assert false
+  else
+    let a = Char.code s.[0] in
+    if len = 1 then
+      a
+    else
+      let b = (Char.code s.[1]) land 0b0011_1111 in
+      if len = 2 then
+        ((a land 0b0001_1111) lsl 6)
+        lor b
+      else
+        let c = (Char.code s.[2]) land 0b0011_1111 in
+        if len = 3 then
+          ((a land 0b0000_1111) lsl 12)
+          lor (b lsl 6)
+          lor c
+        else
+          if len = 4 then
+            let d = (Char.code s.[3]) land 0b0011_1111 in
+            ((a land 0b0000_0111) lsl 18)
+            lor (b lsl 12)
+            lor (c lsl 6)
+            lor d
+          else
+            assert false
+
+let opt_of_char c =
+  match c with
+  | 'i' -> Caseless
+  | 'J' -> Allow_duplicate_names
+  | 'm' -> Multiline
+  | 's' -> Dotall
+  | 'U' -> Default_lazy
+  | 'x' -> Ignore_whitespace
+  | _ -> assert false
+}
+
+let line = [^'\n']* '\n'?
+let alnum = ['A'-'Z' 'a'-'z' '0'-'9']
+let digit = ['0'-'9']
+let xdigit = ['A'-'F' 'a'-'f' '0'-'9']
+let nonneg = ['0'-'9']+
+let group_name = ['A'-'Z' 'a'-'z' '_'] ['A'-'Z' 'a'-'z' '_' '0'-'9']*
+
+let subroutine_name = (* TODO: is this correct? *)
+  ['A'-'Z' 'a'-'z' '_'] ['A'-'Z' 'a'-'z' '_' '0'-'9']*
+
+let ascii = ['\000'-'\127']
+
+(* 110xxxxx *)
+let utf8_head_byte1 = ['\192'-'\223']
+
+(* 1110xxxx *)
+let utf8_head_byte2 = ['\224'-'\239']
+
+(* 11110xxx *)
+let utf8_head_byte3 = ['\240'-'\247']
+
+(* 10xxxxxx *)
+let utf8_tail_byte = ['\128'-'\191']
+
+let utf8 = ascii
+         | utf8_head_byte1 utf8_tail_byte
+         | utf8_head_byte2 utf8_tail_byte utf8_tail_byte
+         | utf8_head_byte3 utf8_tail_byte utf8_tail_byte utf8_tail_byte
+
+let unicode_character_property = ['A'-'Z'] alnum*
+
+(* This is for the extended mode, which ignores whitespace like in usual
+   programming languages.
+   TODO: check if non-ascii whitespace should also be ignored in that mode
+   (it would be annoying to implement but doable).
+*)
+let ascii_whitespace =  ['\t' '\n' '\011' '\012' '\r' ' ']
+
+let btk_name = [^')']*
+
+rule token conf = parse
+  (* Some of these are only allowed at the beginning of the input.
+     For now, we allow them everywhere and ignore them if possible. *)
+  | "(*LIMIT_MATCH=" digit+ ")" { token conf lexbuf }
+  | "(*LIMIT_RECURSION=" digit+ ")" { token conf lexbuf }
+  | "(*NO_AUTO_POSSESS)" { token conf lexbuf }
+  | "(*NO_START_OPT)" { token conf lexbuf }
+  | "(*UTF8)" { token conf lexbuf }
+  | "(*UTF16)" { (* unsupported; should fail? *) token conf lexbuf }
```

Comment:
I don't think we should fail.

Reply:
I agree! I think I was tired and confused the encoding of the regexp (must be utf-8) itself with the encoding of the input of the regexp (could be anything).